File: search-tutorial/typescript/search-basic.ts
/**
 * Gloo AI Search API - Basic TypeScript Example
 *
 * This script demonstrates how to use the Gloo AI Search API
 * to perform semantic search on your ingested content.
 */

import axios from "axios";
import dotenv from "dotenv";
import { TokenManager, validateCredentials } from "./auth";

dotenv.config();

// --- Configuration ---
const CLIENT_ID = process.env.GLOO_CLIENT_ID || "YOUR_CLIENT_ID";
const CLIENT_SECRET = process.env.GLOO_CLIENT_SECRET || "YOUR_CLIENT_SECRET";
const TENANT = process.env.GLOO_TENANT || "your-tenant-name";

const TOKEN_URL = "https://platform.ai.gloo.com/oauth2/token";
const SEARCH_URL = "https://platform.ai.gloo.com/ai/data/v1/search";

// --- Types ---
interface SearchMetadata {
  distance: number;
  certainty: number;
  score: number;
  creation_time: string;
  last_update_time: string;
}

interface SearchProperties {
  item_title: string;
  type: string;
  author: string[];
  snippet: string;
  summaries?: Record<string, string>;
  biblical_analysis?: Record<string, unknown>;
  [key: string]: unknown;
}

interface SearchResult {
  uuid: string;
  metadata: SearchMetadata;
  properties: SearchProperties;
  collection: string;
}

interface SearchResponse {
  data: SearchResult[];
  intent: number;
}

export class SearchClient {
  private tokenManager: TokenManager;

  constructor(tokenManager: TokenManager) {
    this.tokenManager = tokenManager;
  }

  async search(query: string, limit: number = 10): Promise<SearchResponse> {
    const token = await this.tokenManager.ensureValidToken();

    const payload = {
      query,
      collection: "GlooProd",
      tenant: TENANT,
      limit,
      certainty: 0.5,
    };

    try {
      const response = await axios.post<SearchResponse>(SEARCH_URL, payload, {
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json",
        },
        timeout: 60000,
      });
      return response.data;
    } catch (error: any) {
      if (error.response) {
        console.error(
          `Search failed with status ${error.response.status}: ${JSON.stringify(error.response.data)}`
        );
      }
      throw new Error(`Search request failed: ${error.message}`);
    }
  }
}

async function basicSearch(query: string, limit: number = 10): Promise<void> {
  const tokenManager = new TokenManager(CLIENT_ID, CLIENT_SECRET, TOKEN_URL);
  const searchClient = new SearchClient(tokenManager);

  console.log(`Searching for: '${query}'`);
  console.log(`Limit: ${limit} results\n`);

  const results = await searchClient.search(query, limit);

  if (!results.data || results.data.length === 0) {
    console.log("No results found.");
    return;
  }

  console.log(`Found ${results.data.length} results:\n`);

  results.data.forEach((result, i) => {
    const props = result.properties;
    const metadata = result.metadata;

    console.log(`--- Result ${i + 1} ---`);
    console.log(`Title: ${props.item_title || "N/A"}`);
    console.log(`Type: ${props.type || "N/A"}`);
    console.log(`Author: ${(props.author || ["N/A"]).join(", ")}`);
    console.log(`Relevance Score: ${(metadata.certainty || 0).toFixed(4)}`);

    if (props.snippet) {
      console.log(`Snippet: ${props.snippet.substring(0, 200)}...`);
    }

    console.log();
  });
}

function printUsage(): void {
  console.log("Usage:");
  console.log("  npx ts-node search-basic.ts <query> [limit]");
  console.log();
  console.log("Examples:");
  console.log('  npx ts-node search-basic.ts "How can I know my purpose?"');
  console.log('  npx ts-node search-basic.ts "purpose" 5');
}

async function main(): Promise<void> {
  validateCredentials(CLIENT_ID, CLIENT_SECRET);

  const args = process.argv.slice(2);

  if (args.length < 1) {
    printUsage();
    process.exit(1);
  }

  const query = args[0];
  const limit = args[1] ? parseInt(args[1], 10) : 10;

  try {
    await basicSearch(query, limit);
  } catch (error: any) {
    console.error(`An error occurred: ${error.message}`);
    process.exit(1);
  }
}

// Export for use by server
export { TOKEN_URL, SEARCH_URL, CLIENT_ID, CLIENT_SECRET, TENANT };

// Run if executed directly
if (require.main === module) {
  main();
}

File: search-tutorial/python/search_basic.py
#!/usr/bin/env python3
"""
Gloo AI Search API - Basic Python Example

This script demonstrates how to use the Gloo AI Search API
to perform semantic search on your ingested content.
"""

import requests
import sys
from typing import Dict, Any
from auth import TokenManager, validate_credentials
from config import CLIENT_ID, CLIENT_SECRET, TENANT, TOKEN_URL, SEARCH_URL
from utils import normalize_limit


class SearchClient:
    """Handles search requests to the Gloo AI Search API."""

    def __init__(self, token_manager: TokenManager):
        self.token_manager = token_manager

    def search(self, query: str, limit: int = 10) -> Dict[str, Any]:
        """
        Perform a semantic search query.

        Args:
            query: The search query string
            limit: Maximum number of results to return (10-100 recommended)

        Returns:
            Dictionary containing search results
        """
        token = self.token_manager.ensure_valid_token()
        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json"
        }

        payload = {
            "query": query,
            "collection": "GlooProd",
            "tenant": TENANT,
            "limit": limit,
            "certainty": 0.5
        }

        try:
            response = requests.post(
                SEARCH_URL,
                headers=headers,
                json=payload,
                timeout=60
            )
            response.raise_for_status()
            return response.json()

        except requests.exceptions.RequestException as e:
            if hasattr(e, 'response') and e.response is not None:
                print(f"Search failed with status {e.response.status_code}: {e.response.text}")
            raise Exception(f"Search request failed: {e}")


class SearchApp:
    """Main application class for search operations."""

    def __init__(self):
        self.token_manager = TokenManager(CLIENT_ID, CLIENT_SECRET, TOKEN_URL)
        self.search_client = SearchClient(self.token_manager)

    def basic_search(self, query: str, limit: int = 10) -> None:
        """
        Perform a basic search and display results.

        Args:
            query: The search query
            limit: Number of results to return
        """
        try:
            print(f"Searching for: '{query}'")
            print(f"Limit: {limit} results\n")

            results = self.search_client.search(query, limit)

            if not results.get('data'):
                print("No results found.")
                return

            print(f"Found {len(results['data'])} results:\n")

            for i, result in enumerate(results['data'], 1):
                props = result.get('properties', {})
                metadata = result.get('metadata', {})

                print(f"--- Result {i} ---")
                print(f"Title: {props.get('item_title', 'N/A')}")
                print(f"Type: {props.get('type', 'N/A')}")
                print(f"Author: {', '.join(props.get('author', ['N/A']))}")
                print(f"Relevance Score: {metadata.get('certainty', 0):.4f}")

                # Display snippet (first 200 chars)
                snippet = props.get('snippet', '')
                if snippet:
                    print(f"Snippet: {snippet[:200]}...")

                print()

        except Exception as e:
            print(f"Search failed: {e}")
            sys.exit(1)

    def print_usage(self) -> None:
        """Print usage information."""
        print("Usage:")
        print("  python search_basic.py <query> [limit]")
        print()
        print("Arguments:")
        print("  query   Search query string (required)")
        print("  limit   Number of results to return (optional, default: 10)")
        print()
        print("Examples:")
        print('  python search_basic.py "How can I know my purpose?"')
        print('  python search_basic.py "purpose" 5')


def main():
    """Main entry point."""
    validate_credentials(CLIENT_ID, CLIENT_SECRET)
    app = SearchApp()

    if len(sys.argv) < 2:
        app.print_usage()
        sys.exit(1)

    query = sys.argv[1]
    limit = normalize_limit(sys.argv[2], 10) if len(sys.argv) > 2 else 10

    try:
        app.basic_search(query, limit)
    except KeyboardInterrupt:
        print("\nSearch cancelled by user")
        sys.exit(0)
    except Exception as e:
        print(f"An error occurred: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
